ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/DebugDataDumper.java

```diff
@@ -0,0 +1,208 @@
+    createDirectory(this.directory.resolve(INVALID_BLOCK_DIR), INVALID_BLOCK_DIR, "invalid blocks");
+  }
+
+  public void saveGossipMessageDecodingError(
+      final String topic,
+      final Optional<UInt64> arrivalTimestamp,
+      final Bytes originalMessage,
+      final Throwable error) {
+    if (!enabled) {
+      return;
+    }
+    final String formattedTimestamp = formatOptionalTimestamp(arrivalTimestamp);
+    final String fileName = String.format("%s.ssz", formattedTimestamp);
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR)
+            .resolve(DECODING_ERROR_SUB_DIR)
+            .resolve(topic.replaceAll("/", "_"));
+    final boolean success =
+        saveBytesToFile(
+            "gossip message with decoding error", topicPath.resolve(fileName), originalMessage);
+    if (success) {
+      LOG.warn("Failed to decode gossip message on topic {}", topic, error);
+    }
+  }
+
+  public void saveGossipRejectedMessageToFile(
+      final String topic,
+      final Optional<UInt64> arrivalTimestamp,
+      final Bytes decodedMessage,
+      final Optional<String> reason) {
+    if (!enabled) {
+      return;
+    }
+    final String formattedTimestamp = formatOptionalTimestamp(arrivalTimestamp);
+    final String fileName = String.format("%s.ssz", formattedTimestamp);
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR).resolve(REJECTED_SUB_DIR).resolve(topic.replaceAll("/", "_"));
+    final boolean success =
+        saveBytesToFile("rejected gossip message", topicPath.resolve(fileName), decodedMessage);
+    if (success) {
+      LOG.warn(
+          "Rejecting gossip message on topic {}, reason: {}",
+          topic,
+          reason.orElse("failed validation"));
+    }
+  }
+
+  public void saveInvalidBlockToFile(
+      final UInt64 slot,
+      final Bytes32 blockRoot,
+      final Bytes blockSsz,
+      final String failureReason,
+      final Optional<Throwable> failureCause) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s_%s.ssz", slot, blockRoot.toUnprefixedHexString());
+    final boolean success =
+        saveBytesToFile("invalid block", Path.of(INVALID_BLOCK_DIR).resolve(fileName), blockSsz);
+    if (success) {
+      LOG.warn(
+          "Rejecting invalid block at slot {} with root {} because {}",
+          slot,
+          blockRoot,
+          failureReason,
+          failureCause.orElse(null));
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean saveBytesToFile(
+      final String description, final Path relativeFilePath, final Bytes bytes) {
+    final Path path = directory.resolve(relativeFilePath);
+    try {
+      Files.write(path, bytes.toArray());
+    } catch (NoSuchFileException e) {
+      if (!path.getParent().toFile().mkdirs()) {
+        LOG.error("Failed to save {} bytes to file.", description, e);
+        return false;
+      }
+      return saveAfterCreatingTopicDirectory(description, relativeFilePath, bytes);
```

Comment:
if we're going to build things in this way, lets be a lot more careful about this.
We're possibly throwing exceptions in an exceptionHandler, but we're also just calling a function - lets do `mkdirs` in the child function instead, and then this is a lot less susceptible to failure.